server/api/reviews.go

```diff
@@ -17,8 +23,180 @@ type ReviewBody struct {
 	Categories []string `json:"categories"`
 }
 
-func getAll() ([]*db.Review, error) {
-	return utils.FilterReviews(bson.D{{}})
+func getAll(sortQuery string, sortOrder string, searchQuery string, categoriesQuery string) ([]db.Review, error) {
+	var reviewsCursor *mongo.Cursor
+	var err error
+
+	DB, err := db.DB()
+
+	var reviews []db.Review
+
+	// close db connection
+	defer func() {
+		if err := DB.Client().Disconnect(db.Ctx); err != nil {
+			return
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+
+	opts := options.Find()
+
+	if len(sortQuery) > 0 {
+
+		if len(sortOrder) < 0 {
+			return nil, errors.New("Specify your sorting order!")
+		}
+
+		recencySort, err := utils.ConvertStringToNum(sortOrder)
+
+		if err != nil {
+			return nil, errors.New("Invalid value for sort operation")
+		}
+
+		if sortQuery != "likes" && sortQuery != "book_name" {
+			return nil, errors.New("Invalid sort Query")
+		}
+
+		opts.SetSort(bson.D{{Key: sortQuery, Value: recencySort}})
+
+	} else if len(searchQuery) == 0 {
+		opts.SetSort(bson.D{{Key: "created_at", Value: -1}})
+	}
+
+	categoriesArr := strings.Split(categoriesQuery, ",")
+
+	categoriesFilter := bson.M{"categories": bson.M{"$in": categoriesArr}}
+
+	if len(searchQuery) > 0 && len(categoriesQuery) > 0 {
+		reviewsCursor, err = DB.Collection(db.ReviewsCollection).Find(db.Ctx, categoriesFilter, opts)
+		return search(reviewsCursor, err, searchQuery)
+	} else if len(searchQuery) > 0 {
+		reviewsCursor, err = DB.Collection(db.ReviewsCollection).Find(db.Ctx, bson.M{}, opts)
+		return search(reviewsCursor, err, searchQuery)
+	} else if len(categoriesQuery) > 0 {
+		reviewsCursor, err = DB.Collection(db.ReviewsCollection).Find(db.Ctx, categoriesFilter, opts)
+	} else {
+		reviewsCursor, err = DB.Collection(db.ReviewsCollection).Find(db.Ctx, bson.M{}, opts)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err = reviewsCursor.All(db.Ctx, &reviews); err != nil {
+		return nil, err
+	}
+
+	return reviews, nil
+}
+
+func search(cursor *mongo.Cursor, err error, search string) ([]db.Review, error) {
+
+	var reviews []db.Review
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err = cursor.All(db.Ctx, &reviews); err != nil {
+		return nil, err
+	}
+
+	searchKeywords := strings.Split(search, " ")
+
+	reviewsWithKeyword := make(map[primitive.ObjectID]db.Review)
+
+	// Review body should contain matched keyword atleast once
+	for _, item := range searchKeywords {
+		for _, entry := range reviews {
+			if contains(strings.Split(strings.ToLower(strings.Trim(entry.Text, " ")), " "), item) {
+				reviewsWithKeyword[entry.ID] = entry
+			}
+		}
+	}
+
+	withKeywordsReviews := make([]db.Review, 0)
+
+	for _, value := range reviewsWithKeyword {
+		withKeywordsReviews = append(withKeywordsReviews, value)
+	}
+
+	occurences := make(map[primitive.ObjectID]map[string]int)
+
+	//Find occurence of keyword in each review
```

Comment:
Instead of keeping occurence of each keyword, actually we only need the `summation of occurrences of all keywords`. So maybe it is easier you just keep `occurences of all keywords` instead of keeping them separately.